src/read/array/binary.rs

```diff
@@ -185,3 +189,106 @@ where
         }
     }
 }
+
+pub fn read_binary<O: Offset, R: NativeReadBuf>(
+    reader: &mut R,
+    is_nullable: bool,
+    data_type: DataType,
+    page_metas: Vec<PageMeta>,
+) -> Result<Box<dyn Array>> {
+    let num_values = page_metas.iter().map(|p| p.num_values as usize).sum();
+
+    let total_length: usize = page_metas.iter().map(|p| p.length as usize).sum();
+
+    let mut off_offset = 0;
+    let mut buf_offset = 0;
+    let mut validity_builder = if is_nullable {
+        Some(MutableBitmap::with_capacity(num_values))
+    } else {
+        None
+    };
+    let mut scratch = vec![];
+    let out_off_len = num_values + 2;
+    // don't know how much space is needed for the buffer,
+    // if not enough, it may need to be reallocated several times.
+    let out_buf_len = total_length * 4;
+    let mut out_offsets: Vec<O> = Vec::with_capacity(out_off_len);
+    let mut out_buffer: Vec<u8> = Vec::with_capacity(out_buf_len);
+    for page_meta in page_metas {
+        let length = page_meta.num_values as usize;
+        if let Some(ref mut validity_builder) = validity_builder {
+            read_validity(reader, length, validity_builder)?;
+        }
+        let start_offset = out_offsets.last().copied();
+        batch_read_buffer(
+            reader,
+            off_offset,
+            length + 1,
+            &mut scratch,
+            &mut out_offsets,
```

Comment:
why not back-forward the `out_offset` as `&mut out_offsets[out_offsets.len() - 1 ..]`